### PR TITLE
DWARF processing fixes

### PR DESCRIPTION
**Pull request by erozenfeld**
Add support for binaries with both .debug_ranges and .debug_rnglists sections .debug_rnglists was added in DWARF5 to replace .debug_ranges. This changes handles cases when the binary has compilation units with DWARF5 info and compilation units with pre-version-5 DWARF info.

Fix processing of DW_AT_ranges attributes to allow DW_FORM_data4 and DW_FORM_data8 This affects DWARF pre-version-4 and was broken by https://github.com/google/autofdo/commit/dd1395a351d63e568c43429e467a547541c66553 .

Fixes https://github.com/google/autofdo/issues/154

### CODE REVIEW

**Comment by erozenfeld** [comment]
@snehasish @shenhanc78 Please take a look.

**Review by snehasish — COMMENTED** [review]
Thanks, would it be possible to add a test for this?

**Comment by erozenfeld** [comment]
> Thanks, would it be possible to add a test for this?

I'm happy to do that but currently no GCC tests run in CI. Is it possible to enable them?

**Comment by snehasish** [comment]
> Is it possible to enable them?

Sure, can you send a separate PR to do so?

**Comment by erozenfeld** [comment]
@snehasish  @shenhanc78 I see this comment in https://github.com/google/autofdo/pull/213#issue-2414887845:

``. We also build GCOV tools built with GCC but do not test them yet since a couple of tests are failing.``

How can I run those tests and repro the failures?

**Comment by snehasish** [comment]
Looks like there is just one failure in the symbolize test. I created draft PR #225 which adds make test to the gcov build. Here is the log: https://github.com/google/autofdo/actions/runs/10359876746/job/28677175564?pr=225

**Comment by erozenfeld** [comment]
All of these tests are from third_party directories, e.g., symbolize test is from `third_party\glog\src\symbolize_unittest.cc`. `create_gcov` doesn't use `third_party\glog\src\symbolize` and that test passes for me locally. Are there any tests that test `create_gcov` or anything it uses?

**Comment by snehasish** [comment]
That's a good point, I don't have any experience on the gcov side but taking a quick look at the source, it doesn't look like we have any tests that cover gcov profile generation. Maybe we can skip the symbolize test on the CI when built for gcov, enable the CI for gcov and then add a test which covers this PR too?

**Comment by erozenfeld** [comment]
I can't repro the symbolize test failure. Do you know how to disable it in CI when built for gcov?

**Comment by erozenfeld** [comment]
@snehasish I opened https://github.com/google/autofdo/pull/226. I disabled third_party tests similar to how it's done for build-llvm. I have a test for this PR, will update it once https://github.com/google/autofdo/pull/226 is merged.

**Comment by erozenfeld** [comment]
@snehasish I added a test for this fix. PTAL

File: legacy_addr2line.cc [hunk 1]
```diff
@@ -105,27 +105,28 @@ bool Google3Addr2line::Prepare() {
...
-  if (debug_ranges_data == NULL) {
-      GetSection(sections, ".debug_rnglists", &debug_ranges_data, &debug_ranges_size);
-      if (debug_ranges_data != NULL) 
-        is_rnglists_section = true;
-      else 
-        LOG(WARNING) << "File '" << binary_name_ << "' does not have .debug_ranges nor .debug_rnglists sections.";
+  GetSection(sections, ".debug_rnglists", &debug_rnglists_data, &debug_rnglists_size);
+
+  if (debug_ranges_data == NULL && debug_rnglists_data == NULL) {
+    LOG(WARNING) << "File '" << binary_name_ << "' does not have .debug_ranges nor .debug_rnglists sections.";
```
Comment (snehasish) [inline legacy_addr2line.cc:122]: I'm not sure if we can proceed if both of them are not available for newer versions of dwarf. Should this be a fatal error instead (based on the dwarf version)?

File: addr2line_test.cc [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#include "addr2line.h"
+#include "third_party/abseil/absl/strings/str_cat.h"
+#include "util/testing/status_matchers.h"
+
+namespace {
+
+using ::devtools_crosstool_autofdo::Addr2line;
+
+TEST(Addr2lineTest, ComdatFuncHasNoDwp) {
```
Comment (snehasish) [inline addr2line_test.cc:9]: Should we change the name of the test? I don't think this has anything to do with Comdats :)

**Review by snehasish — APPROVED** [review]
lgtm with a couple of minor comments.

File: legacy_addr2line.cc (hunk 1 above)
Comment (erozenfeld) [reply legacy_addr2line.cc:122]: It's possible that the binary has no address ranges in its dwarf info. If we attempt to use address ranges later, we will assert at that time (in dwarf3ranges.cc).

File: addr2line_test.cc (hunk 2 above)
Comment (erozenfeld) [reply addr2line_test.cc]: Good catch, fixed.

**Pull request merged**